File: backend/kinesisSimulator.js
// ============================================================
// kinesisSimulator.js — AWS Kinesis Data Stream Simulator
// ============================================================
// Simulates a Kinesis stream with shards. Raw weather data is
// put into the stream and consumers (Lambda) are notified.
// ============================================================

const EventEmitter = require('events');

class KinesisSimulator extends EventEmitter {
    constructor(streamName = 'WeatherDataStream', shardCount = 1) {
        super();
        this.streamName = streamName;
        this.shardCount = shardCount;
        this.shards = Array.from({ length: shardCount }, (_, i) => ({
            shardId: `shard-${i}`,
            records: [],
            sequenceNumber: 0
        }));
        this.status = 'ACTIVE';
        this.createdAt = new Date().toISOString();
        this.totalRecordsIngested = 0;

        console.log(`[Kinesis] Stream "${this.streamName}" initialized with ${this.shardCount} shard(s)`);
    }

    /**
     * Put a record into the stream (routes to a shard based on partition key).
     */
    putRecord(data, partitionKey) {
        const shardIndex = this._hashPartitionKey(partitionKey) % this.shardCount;
        const shard = this.shards[shardIndex];

        const record = {
            sequenceNumber: String(++shard.sequenceNumber).padStart(21, '0'),
            partitionKey,
            data,
            approximateArrivalTimestamp: new Date().toISOString(),
            shardId: shard.shardId
        };

        shard.records.push(record);
        this.totalRecordsIngested++;

        // Keep shard buffer bounded
        if (shard.records.length > 1000) {
            shard.records = shard.records.slice(-500);
        }

        // Emit for consumers (Lambda)
        this.emit('record', record);

        console.log(`[Kinesis] Record ingested → shard: ${shard.shardId}, seq: ${record.sequenceNumber}, key: ${partitionKey}`);

        return {
            shardId: shard.shardId,
            sequenceNumber: record.sequenceNumber
        };
    }

    /**
     * Put multiple records at once.
     */
    putRecords(records) {
        return records.map(({ data, partitionKey }) => this.putRecord(data, partitionKey));
    }

    /**
     * Get stream description (like AWS DescribeStream).
     */
    describeStream() {
        return {
            streamName: this.streamName,
            streamStatus: this.status,
            shardCount: this.shardCount,
            totalRecordsIngested: this.totalRecordsIngested,
            createdAt: this.createdAt,
            shards: this.shards.map(s => ({
                shardId: s.shardId,
                recordCount: s.records.length,
                latestSequenceNumber: s.sequenceNumber
            }))
        };
    }

    /**
     * Simple hash of partition key to determine shard.
     */
    _hashPartitionKey(key) {
        let hash = 0;
        for (let i = 0; i < key.length; i++) {
            hash = ((hash << 5) - hash) + key.charCodeAt(i);
            hash |= 0;
        }
        return Math.abs(hash);
    }
}

module.exports = KinesisSimulator;

File: backend/kinesis_simulator.py
# ============================================================
# kinesis_simulator.py — AWS Kinesis Stream Simulator
# ============================================================
# Simulates an AWS Kinesis data stream with shards.
# Records are distributed across shards based on partition key.
# Emits events when records are added.
# ============================================================

import hashlib
from datetime import datetime
from typing import Callable, List


class KinesisSimulator:
    """Simulates an AWS Kinesis stream with shards."""
    
    def __init__(self, stream_name: str, shard_count: int = 2):
        self.stream_name = stream_name
        self.shard_count = shard_count
        self.shards = {f'shard-{i}': [] for i in range(shard_count)}
        self.record_count = 0
        self.listeners: List[Callable] = []
        self.created_at = datetime.now().isoformat()
        
        print(f'[Kinesis] ✓ Stream created: {stream_name} (shards: {shard_count})')
    
    def put_record(self, data: dict, partition_key: str) -> dict:
        """Put a record into the stream."""
        # Hash partition key to determine shard
        shard_id = self._get_shard_for_key(partition_key)
        
        record = {
            'sequenceNumber': str(self.record_count),
            'partitionKey': partition_key,
            'data': data,
            'approximateArrivalTimestamp': datetime.now().isoformat()
        }
        
        self.shards[shard_id].append(record)
        self.record_count += 1
        
        # Notify listeners
        self._emit('record', record)
        
        # Cap shard size at 1000 records
        if len(self.shards[shard_id]) > 1000:
            self.shards[shard_id].pop(0)
        
        return {
            'shardId': shard_id,
            'sequenceNumber': record['sequenceNumber']
        }
    
    def _get_shard_for_key(self, partition_key: str) -> str:
        """Determine which shard to use based on partition key."""
        hash_val = int(hashlib.md5(partition_key.encode()).hexdigest(), 16)
        shard_index = hash_val % self.shard_count
        return f'shard-{shard_index}'
    
    def on(self, event: str, callback: Callable) -> None:
        """Register an event listener."""
        if event == 'record':
            self.listeners.append(callback)
    
    def _emit(self, event: str, data: dict) -> None:
        """Emit an event to all listeners."""
        for listener in self.listeners:
            try:
                listener(data)
            except Exception as e:
                print(f'[Kinesis] Listener error: {e}')
    
    def describe_stream(self) -> dict:
        """Return stream metadata."""
        created_dt = datetime.fromisoformat(self.created_at)
        uptime_ms = int((datetime.now() - created_dt).total_seconds() * 1000)
        
        return {
            'streamName': self.stream_name,
            'streamStatus': 'ACTIVE',
            'shardCount': self.shard_count,
            'recordCount': self.record_count,
            'shards': {
                shard_id: len(records) 
                for shard_id, records in self.shards.items()
            },
            'createdAt': self.created_at,
            'uptime': uptime_ms
        }
